chain/solana/src/chain.rs

```diff
@@ -0,0 +1,299 @@
+        };
+
+        let logger = self
+            .logger_factory
+            .subgraph_logger(&deployment)
+            .new(o!("component" => "FirehoseBlockStream"));
+
+        let firehose_mapper = Arc::new(FirehoseMapper {});
+        let firehose_cursor = store.block_cursor();
+
+        Ok(Box::new(FirehoseBlockStream::new(
+            firehose_endpoint,
+            firehose_cursor,
+            firehose_mapper,
+            adapter,
+            filter,
+            start_blocks,
+            logger,
+        )))
+    }
+
+    async fn new_polling_block_stream(
+        &self,
+        _deployment: DeploymentLocator,
+        _writable: Arc<dyn WritableStore>,
+        _start_blocks: Vec<BlockNumber>,
+        _subgraph_start_block: Option<BlockPtr>,
+        _filter: Arc<Self::TriggerFilter>,
+        _metrics: Arc<BlockStreamMetrics>,
+        _unified_api_version: UnifiedMappingApiVersion,
+    ) -> Result<Box<dyn BlockStream<Self>>, Error> {
+        panic!("SOLANA does not support polling block stream")
+    }
+
+    fn chain_store(&self) -> Arc<dyn ChainStore> {
+        self.chain_store.clone()
+    }
+
+    async fn block_pointer_from_number(
+        &self,
+        _logger: &Logger,
+        _number: BlockNumber,
+    ) -> Result<BlockPtr, IngestorError> {
+        // FIXME (Solana): Hmmm, what to do with this?
+        Ok(BlockPtr {
+            hash: BlockHash::from(vec![0xff; 32]),
+            number: 0,
+        })
+    }
+
+    fn runtime_adapter(&self) -> Arc<Self::RuntimeAdapter> {
+        Arc::new(RuntimeAdapter {})
+    }
+
+    fn is_firehose_supported(&self) -> bool {
+        true
+    }
+}
+
+pub struct TriggersAdapter {}
+
+#[async_trait]
+impl TriggersAdapterTrait<Chain> for TriggersAdapter {
+    fn ancestor_block(
+        &self,
+        _ptr: BlockPtr,
+        _offset: BlockNumber,
+    ) -> Result<Option<codec::Block>, Error> {
+        // FIXME (Solana):  Might not be necessary for Solana support for now
+        Ok(None)
+    }
+
+    async fn scan_triggers(
+        &self,
+        _from: BlockNumber,
+        _to: BlockNumber,
+        _filter: &TriggerFilter,
+    ) -> Result<Vec<BlockWithTriggers<Chain>>, Error> {
+        // FIXME (Solana): Scanning triggers makes little sense in Firehose approach, let's see
+        Ok(vec![])
+    }
+
+    async fn triggers_in_block(
+        &self,
+        _logger: &Logger,
+        block: codec::Block,
+        _filter: &TriggerFilter,
+    ) -> Result<BlockWithTriggers<Chain>, Error> {
+        let shared_block = Arc::new(block.clone());
+        let instructions = block.transactions.iter().flat_map(|transaction| {
+            //let transaction_id = transaction.id.clone();
+            let b = shared_block.clone();
+            let tx = transaction.clone();
+            transaction.instructions.iter().flat_map(move |instruction| {
+                Some(trigger::InstructionWithInfo {
+                    instruction: instruction.clone(),
```

Comment:
I did not check everything, but I found it a bit intense to have the need to fully clone the whole instruction, maybe we could keep only a shared reference to the data (it might not be possible due to how we use it, we will need to investigate that further).

Reply:
This is even more true because those a later filtered, so we pay high copy price to discard 90% of them later. Minimally if we really need to copy, it would be much better to clone after filtering has been done.

Reply:
we can check together if there is a better way.